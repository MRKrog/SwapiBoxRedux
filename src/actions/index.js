
export const saveMovie = (info) => ({
  type: 'SAVE_MOVIE',
  info
})

export const savePeople = (people) => ({
  type: 'SAVE_PEOPLE',
  people
})

export const savePlanets = (planets) => ({
  type: 'SAVE_PLANETS',
  planets
})

export const saveVehicles = (vehicles) => ({
  type: 'SAVE_VEHICLES',
  vehicles
})

export const saveFavAmount = (favorites) => ({
  type: "UPDATE_COUNT",
  favorites
})
