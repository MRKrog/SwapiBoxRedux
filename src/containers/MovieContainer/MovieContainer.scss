
.MovieContainer {
  width: 100%;
  height: 100%;
  padding: 4em;

  .MovieInfoContainer {
    max-width: 1200px;
    position: relative;

    .MovieCardContainer {
      background: #505050bd;
      padding: 2%;
      box-shadow: inset 0px 0px 10px 1px #231f20;
      overflow: scroll;
      min-height: 500px;
      max-height: 500px;
      position: relative;

      .MovieTitleDisplay {
        position: absolute;
        z-index: 9;
        top: -0.5em;
        left: 1em;

        h3 {
          font-family: $font-Label;
          color: #231f2094;
          text-transform: uppercase;
          font-size: 5em;
        }
      }

      .MovieCardsDisplay {
        display: flex;
        flex-wrap: wrap;
        justify-content: flex-start;
        color:#fff;
        z-index: 10;
        position: relative;
        margin-top: 1.5em;
      }

      h3 {
        margin:0;
      }
    }

  }

  .shadowBottom {
    position: relative;
    z-index: 999;

    &::after {
      content:'';
      position: absolute;
      width: 100%;
      bottom: 0px;
      height: 40px;
      left: 0;
      background: linear-gradient(to bottom, rgba(76,76,76,0.1) 0%,rgba(43,43,43,0.75) 72%,rgba(19,19,19,1))
    }

  }

}
