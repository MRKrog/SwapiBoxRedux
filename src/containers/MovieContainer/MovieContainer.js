import React, { Component } from 'react';
import { connect } from 'react-redux'
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';

import { saveFavAmount } from '../../actions'
import Card from '../../components/Card/Card';


export class MovieContainer extends Component {
  constructor(props) {
    super(props);

  }

  handleFavBtn = (name, category) => {
    const favIndex = this.props[category].findIndex(card => card.name === name);
    this.changeFav(this.props[category][favIndex]);
    const totalFavs = this.findAllFavorites();
    this.props.storeFavAmount(totalFavs)
  }

  findAllFavorites = () => {
    const allFavorites = [...this.props.people, ...this.props.planets, ...this.props.vehicles]

    const favCards = allFavorites.filter(card => {
      return card.favorite === true;
    })
    console.log(favCards);
    return favCards.length
  }

  changeFav = (card) => {
    if(!card.favorite) {
      card.favorite = true
    } else {
      card.favorite = false
    }
  }


  render() {
    const { id } = this.props;
    let btnActive = 'btnActive';

    const cardsToDisplay = this.props.cardData.map((info, index) => (
      <Card info={info}
            handleFavBtn={this.handleFavBtn}
            key={index}
      />
    ))

    return (
      <div className="MovieContainer">
        <section className="MovieButtons">
          <NavLink to="/People">
            <button className={id === 'people' ? btnActive : ''} onClick={() => this.props.fetchPeople('people')}>People</button>
          </NavLink>
          <NavLink to="/Planets">
            <button className={id === 'planets' ? btnActive : ''} onClick={() => this.props.fetchPlanets('planets')}>Planets</button>
          </NavLink>
          <NavLink to="/Vehicles">
            <button className={id === 'vehicles' ? btnActive : ''} onClick={() => this.props.fetchVehicles('vehicles')}>Vehicles</button>
          </NavLink>
        </section>
        <div className="MovieInfoContainer">
          <section className="MovieCardContainer">
            <div className="MovieTitleDisplay"><h3>{id}</h3></div>
            <section className="MovieCardsDisplay">
            {cardsToDisplay}
            </section>
          </section>
          <section className="shadowBottom"></section>
        </div>
      </div>
    )
  }
}

MovieContainer.propTypes = {
  movieData: PropTypes.array
}

export const mapStateToProps = (state) => ({
  people: state.people,
  planets: state.planets,
  vehicles: state.vehicles,
  favAmount: state.favoriteTotal,
})

export const mapDispatchToProps = (dispatch) => ({
  storeFavAmount: (favorites) => dispatch(saveFavAmount(favorites))
})

export default connect(mapStateToProps, mapDispatchToProps)(MovieContainer);
