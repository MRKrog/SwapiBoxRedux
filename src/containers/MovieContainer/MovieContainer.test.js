import React from 'react';
import { shallow } from 'enzyme';

import {MovieContainer, mapStateToProps, mapDispatchToProps} from './MovieContainer';

import * as actions from '../../actions'

const mockData = [
  {
    category: "people",
    favorite: false,
    homeworld: "Tatooine",
    language: "Galactic Basic",
    name: "Luke Skywalker",
    population: "200000",
    species: "Human",
  },
  {
    category: "people",
    favorite: false,
    homeworld: "Tatooine",
    language: "n/a",
    name: "C-3PO",
    population: "200000",
    species: "Droid",
  }
]
const category = 'people'
const mockTotalFavs = 10;
const mockPeople = [{name: 'Luke', favorite: true}, {name: 'Han Solo', favorite: false}]
const mockPlanets = [{name: 'Tattoin', favorite: true}, {name: 'Hoth', favorite: false}]
const mockVehicles = [{name: 'Falcon', favorite: true}, {name: 'Tie Fighter', favorite: false}]

describe('MovieContainer', () => {

  describe('MovieContainer Component', () => {
    let wrapper;
    let url;
    let mockStoreFavAmount;

    beforeEach(() => {
        mockStoreFavAmount = jest.fn()
        wrapper = shallow(<MovieContainer id={category}
                                          cardData={mockData}
                                          people={mockPeople}
                                          planets={mockPlanets}
                                          vehicles={mockVehicles}
                                          storeFavAmount={mockStoreFavAmount}
                                />)


    })

    it('should Match the Snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should invoke storeFavAmount when handleFavBtn is invoked', () => {
      wrapper.instance().handleFavBtn('Hoth', 'planets')

      expect(mockStoreFavAmount).toHaveBeenCalled()
    })

    it('should find all the favorite cards and return its length when findAllFavorites is invoked', () => {
      const favTotals = wrapper.instance().findAllFavorites()
      expect(favTotals).toEqual(4)
    })

    it('should change the cards property of favorite to true when changeFav is invoked', () => {
      const mockCard = {
        category: "people",
        favorite: true,
        homeworld: "Tatooine",
      }

      const expectedCard = {
        category: "people",
        favorite: false,
        homeworld: "Tatooine",
      }

      wrapper.instance().changeFav(mockCard)
      expect(mockCard).toEqual(expectedCard)
    })

  })

  describe('mapStateToProps', () => {

    it('should return an object with the correct properties', () => {
      const mockData = {
        movie: [],
        people: [],
        planets: [],
        vehicles: [],
        favAmount: undefined,
      }
      const expected = {
        people: [],
        planets: [],
        vehicles: [],
        favAmount: undefined,
      }

      const mockProps = mapStateToProps(mockData);
      expect(mockProps).toEqual(expected);
    })

  })

  describe('mapDispatchToProps', () => {

    it('should call dispatch storeMovie', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = actions.saveFavAmount(mockTotalFavs)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.storeFavAmount(mockTotalFavs)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

  })

})
