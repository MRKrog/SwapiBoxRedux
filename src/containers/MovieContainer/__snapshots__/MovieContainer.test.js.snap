// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContainer MovieContainer Component should Match the Snapshot 1`] = `
<div
  className="MovieContainer"
>
  <section
    className="MovieButtons"
  >
    <NavLink
      to="/People"
    >
      <button
        className="btnActive"
        onClick={[Function]}
      >
        People
      </button>
    </NavLink>
    <NavLink
      to="/Planets"
    >
      <button
        className=""
        onClick={[Function]}
      >
        Planets
      </button>
    </NavLink>
    <NavLink
      to="/Vehicles"
    >
      <button
        className=""
        onClick={[Function]}
      >
        Vehicles
      </button>
    </NavLink>
  </section>
  <div
    className="MovieInfoContainer"
  >
    <section
      className="MovieCardContainer"
    >
      <div
        className="MovieTitleDisplay"
      >
        <h3>
          people
        </h3>
      </div>
      <section
        className="MovieCardsDisplay"
      >
        <Card
          handleFavBtn={[Function]}
          info={
            Object {
              "category": "people",
              "favorite": false,
              "homeworld": "Tatooine",
              "language": "Galactic Basic",
              "name": "Luke Skywalker",
              "population": "200000",
              "species": "Human",
            }
          }
          key="0"
        />
        <Card
          handleFavBtn={[Function]}
          info={
            Object {
              "category": "people",
              "favorite": false,
              "homeworld": "Tatooine",
              "language": "n/a",
              "name": "C-3PO",
              "population": "200000",
              "species": "Droid",
            }
          }
          key="1"
        />
      </section>
    </section>
    <section
      className="shadowBottom"
    />
  </div>
</div>
`;
